zbus: Add Zephyr message bus

rodrigopex opened the pull request:
### Pull request for adding zbus to Zephyr.

The Zephyr message bus - Zbus is a lightweight and flexible message bus enabling a simple way for threads to talk to one another. Threads can broadcast messages to all interested observers using zbus. Many-to-many communication is possible. The bus implements message-passing and publish/subscribe communication paradigms that enable threads to communicate synchronously or asynchronously through shared memory. The communication through zbus channel-based, where threads publish and read to and from using messages. Additionally, threads can observe channels and receive notifications when the channels are modified.

### Note for reviewers - Remaining compliance warnings :warning: 
The only one I added is the _ZBUS_ASSERT macro which is necessary to verify the asserts coverage during tests. I have discussed it with @yperess about it. The remaining compliance errors seem to be related to `checkpatch` and/or `clang-format` issues. 

```console
-:517: WARNING:MACRO_WITH_FLOW_CONTROL: Macros with flow control statements should be avoided
#517: FILE: include/zephyr/zbus/zbus.h:104:
+#define _ZBUS_ASSERT(_cond, _fmt, ...)                                                             \
+	({                                                                                         \
+		if (!(_cond)) {                                                                    \
+			LOG_ERR(_fmt, ##__VA_ARGS__);                                              \
+			return -EFAULT;                                                            \
+		}                                                                                  \
+	})

-:602: WARNING:SUSPECT_CODE_INDENT: suspect code indent for conditional statements (8, 8)
#602: FILE: include/zephyr/zbus/zbus.h:189:
+	FOR_EACH_NONEMPTY_TERM(_ZBUS_OBS_EXTERN, (;), _observers)                                 \
+	_ZBUS_STRUCT_DECLARE(zbus_channel, _name) = {                                              \

-:604: WARNING:LONG_LINE: line length of 110 exceeds 100 columns
#604: FILE: include/zephyr/zbus/zbus.h:191:
+		.name = STRINGIFY(_name), .read_only = (_read_only),                                         \
```

### [Zbus presentation slides](https://docs.google.com/presentation/d/1WA_tw0HOz832TGa5v2yQmPzl0M5vSryZ17DDg20hd_4/edit?usp=sharing)

### [Zbus documentation](https://builds.zephyrproject.io/zephyr/pr/48509/docs/services/zbus/index.html)

#### Features table

Characteristic | zbus
-- | --
Made for | Increase code quality by enabling reuse and increasing testability
Metaphor | Messaging bus
Message definition time | Compile-time
Message definition approach | Decentralized, developers can define channels in different files
Message allocation style | Static (compile-time). It is possible to use dynamically allocated messages. See dyn_channel sample 
Message persistency | Persistent, it still exists after processed
Message distribution pattern | Hybrid: Message-passing (listeners) and Publish/subscribe (subscribers)
Subscription time | Compile-time (it can be disabled in execution time by masking the subscriber) and runtime registration is possible
Message transmission style | Direct transmission when using callbacks style. Two-factor for asynchronous transmission where, first, the event dispatcher sends the id of the changed channel to the subscriber, and second the subscriber decides if it reads the content or not. The transmission order is defined by the position of the subscriber on the subscribing list
Observer execution styles | Synchronous for listeners (by callback), asynchronous for subscribers (by queue)
Event dispatcher | There is no centralized event dispatcher. The dispatcher logic is executed by the publisher thread context. It enables the execution in different priority contexts and avoids preemption of the centralized approach
Implementation approach | Static memory, mutexes, sys_slists, and queues
Use code generation (tools/scripts) | No, only macros
Extension mechanism (you can add your functionality) | Yes (see uart_bridge and remote_mock samples)
Maturity | Feature-complete

This pull request aims to add zbus (message bus/event manager) discussed at RFC #45910 to Zephyr. With the implemented tests, it was possible to reach 90%+ of line and 80%+ of branch coverage. All the APIs were designed to be like the Zephyr's. For example, the channel publish is almost the same as a message queue put function. There is a claim/finish like the ring buffer and so on. I guess zbus would help several developers to create even more reusable and decoupled code. The pull request commits are separated into feature code, tests, and samples. I have made individual commits to help reviewers.

Base feature code commit:
- zbus: zbus: Add message bus subsystem to Zephyr

Zbus tests:
- tests: zbus: Add zbus unittests
- tests: zbus: Add zbus integration tests
- tests: zbus: Add zbus user_data usage tests
- tests: zbus: Add zbus dynamic channels tests
- tests: zbus: Add zbus runtime observer registration tests

Zbus samples:
- samples: zbus: Add the hello world sample to zbus
- samples: zbus: Add work queue sample
- samples: zbus: Add dynamic channels sample
- samples: zbus: Add the UART bridge sample
- samples: zbus: Add the remote mock sample
- samples: zbus: Add runtime observer registration sample
- samples: zbus: Add benchmark sample 

Activities during the PR:

- [x] Tests coverage (results: Line 93%, Branch 81%);
- [x] Compliance issues, there are some, but I guess they are not related to the submission;
- [x] General documentation;
- [x] Samples documentation.

rodrigopex committed:
zbus: Add message bus subsystem to Zephyr

Add zbus message bus as a Zephyr subsystem. No message bus
or communication abstraction other than the usual (message queues,
mailboxes, etc.) enabled developers to implement event-driven systems in
Zephyr quickly. Zbus would fill that gap by providing the community with
a lightweight and flexible message bus. The implementation tries to be
closest as possible to the existing ones. We use the claim/finish
approach, and the API for publishing and reading channels are similar
in message queues. Zbus is about channels, messages, and observers.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
tests: zbus: Add zbus unittests

Add zbus subsystem test to verify read-only and hard channels.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
tests: zbus: Add zbus integration tests

Add zbus subsystem test to verify threads integration based on usual
interaction, masking observers, and delayed interaction to verify semaphore
and queue event dispatcher timeout.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
tests: zbus: Add zbus user_data usage tests

Add zbus subsystem test to verify the user_data usage does not affect
the normal behavior.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
tests: zbus: Add zbus dynamic channels tests

Add zbus subsystem test to verify the dynamic channel usage.
It demonstrates the use of static and dynamic external messages.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
samples: zbus: Add the hello world sample to zbus

The hello world sample illustrates a simple way of using zbus. There is
a listener and a subscriber interacting each other by using a read-only
and a regular channel. The APIs are called by using the function
and macro variation.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
samples: zbus: Add work queue sample

The work_queue sample illustrates three reaction styles available for
using zbus. The first is a listener that reacts by callback; use it should
for urgent reaction. The second is a listener that responds by callback;
instead of executing the code, it pushes a job to a work queue that will
be executed shortly but not immediately. The last one is the subscriber
that reacts using a queue; use it for an asynchronous reaction where the
developer would like to control the flow.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
samples: zbus: Add dynamic channels sample

The sample illustrates a way of using dynamically allocated messages in
static channels.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
samples: zbus: Add the UART bridge sample

The sample illustrates a way of send all the channel event to the host
by using the UART, a user_data portion, and the claim/finish APIs.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
samples: zbus: Add the remote mock sample

The sample illustrates a way of exchange message with a host mock
running in a Python script. It can be used to improve the testability
and controllability of the tests.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

gmarull commented on subsys/zbus/Kconfig:
```diff
@@ -0,0 +1,51 @@
+
+if ZBUS
+
+config ZBUS_LOG
+  bool "Log for zbus operation details"
+  depends on LOG
+  default n
```
?

gmarull commented on subsys/zbus/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_sources(zbus.c)
```
please create a library (`zephyr_library()`). Also note that it doesn't make sense to add this commit standalone. This means that previous ones introduce a partial feature that can't be compiled.

gmarull commented on subsys/CMakeLists.txt:
```diff
@@ -32,3 +32,4 @@ add_subdirectory_ifdef(CONFIG_DEMAND_PAGING        demand_paging)
 add_subdirectory(modbus)
 add_subdirectory(sd)
 add_subdirectory(rtio)
+add_subdirectory_ifdef(CONFIG_ZBUS                 zbus)
```
same here wrt commit structure

gmarull commented on tests/subsys/zbus/unittests/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
+  unittests.hard_and_readonly_channels:
+    build_only: false
+    tags: zbus
```
platform_allow to native_posix?

gmarull commented on tests/subsys/zbus/unittests/include/zbus_channels.h:
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2022 Rodrigo Peixoto <rodrigopex@gmail.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+ZBUS_CHAN_DEFINE(version,	       /* Name */
+		 false,		       /* On changes only */
+		 true,		       /* Read only */
+		 struct version,       /* Message type */
+		 NULL,		       /* Validator */
+		 ZBUS_OBSERVERS_EMPTY, /* observers */
+		 ZBUS_MSG_INIT(.major = 0, .minor = 1,
+			       .build = 1023) /* Initial value major 0, minor 1, build 1023 */
+)
```
I assume this shows that Zbus channels can only be defined in this single header. Have you thought about using iterable sections?

gmarull commented on tests/subsys/zbus/integration/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+tests:
+  integration.module_interaction_no_error:
+    build_only: false
+    platform_exclude: qemu_leon3 qemu_malta_be
```
why?

gmarull commented on tests/subsys/zbus/dyn_channel/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+cmake_minimum_required(VERSION 3.13.1)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(integration)
+ set ( CMAKE_C_FLAGS "-save-temps")
```
?

gmarull commented on samples/subsys/zbus/hello_world/include/zbus_messages.h:
```diff
@@ -0,0 +1,22 @@
+
+#ifndef _ZBUS_MESSAGES_H_
+#define _ZBUS_MESSAGES_H_
+#include <zephyr.h>
```
this header is not needed here, you just need `<stdint.h>`

gmarull commented on samples/subsys/zbus/hello_world/sample.yaml:
```diff
@@ -0,0 +1,24 @@
+  sample.zbus.hello_world_build:
+    tags: zbus
+    build_only: true
+    platform_allow: hifive1_revb nrf52840dk_nrf52840 
```
is this sample board specific?

gmarull commented on samples/subsys/zbus/hello_world/src/main.c:
```diff
@@ -0,0 +1,70 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "kernel.h"
```
please don't use `""` to include non-local headers.

gmarull reviewed (commented):
Thanks for your contribution, I've done a first pass review. My main concern is on how channels/messages are defined. Could you take a look at iterable sections? I think they would simplify things and would allow for a cleaner solution.

rodrigopex commented:
@gmarull 

> Thanks for your contribution, I've done a first pass review. 

Thank you for the comments.

> My main concern is on how channels/messages are defined. Could you take a look at iterable sections? I think they would simplify things and would allow for a cleaner solution.

I will take a look at that. This is, in fact, a tricky part of the code.

rodrigopex commented:
I have made some heavy changes to the way of generating channels. Now it is using Iterable Sections. The code is cleaner and smaller. I really good improvement. Thanks a lot, @gmarull, for the advice. There is a lot of work to do. I will address that in the next few days.

gmarull commented:
@rodrigopex please make sure CI passes and that commit structure is ok (for example, https://github.com/zephyrproject-rtos/zephyr/pull/48509/commits/9adf191c0da7329a2a88ae6bbd77ea8bf0889b4b misses proper commit description)

rodrigopex commented:
Sorry for that.

rodrigopex commented:
@gmarull The issues I am facing with CI now seem to be something other them my submission. `Checkpatch` and `clang-format` are fighting each other. The warnings below I could not remove:

```
-:455: WARNING:SUSPECT_CODE_INDENT: suspect code indent for conditional statements (8, 8)
#455: FILE: include/zephyr/zbus/zbus.h:163:
+	FOR_EACH_NONEMPTY_TERM(_ZBUS_OBS_EXTERN, (;), _observers)                                 \
+	STRUCT_SECTION_ITERABLE(zbus_channel, _name) = {                                           \

-:457: WARNING:LONG_LINE: line length of 110 exceeds 100 columns
#457: FILE: include/zephyr/zbus/zbus.h:165:
+		.name = STRINGIFY(_name), .flag = {false, _on_changed, _read_only},                          \

- total: 0 errors, 2 warnings, 3718 lines checked
```

The building errors seem related to iterable sections. I could not replicate those locally. 

```
: && ccache /opt/toolchains/zephyr-sdk-0.14.2/xtensa-intel_apl_adsp_zephyr-elf/bin/xtensa-intel_apl_adsp_zephyr-elf-gcc   zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/__w/zephyr/zephyr/twister-out/intel_adsp_cavs15/samples/subsys/zbus/hello_world/sample.zbus.hello_world/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/arch/arch/xtensa/core/startup/libarch__xtensa__core__startup.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/soc/xtensa/intel_adsp/common/libintel_adsp_common.a  zephyr/subsys/zbus/libsubsys__zbus.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/timer/libdrivers__timer.a  modules/xtensa/libmodules_xtensa_hal.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.obj  -L"/opt/toolchains/zephyr-sdk-0.14.2/xtensa-intel_apl_adsp_zephyr-elf/bin/../lib/gcc/xtensa-intel_apl_adsp_zephyr-elf/10.3.0"  -L/__w/zephyr/zephyr/twister-out/intel_adsp_cavs15/samples/subsys/zbus/hello_world/sample.zbus.hello_world/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  zephyr/soc/xtensa/intel_adsp/common/libintel_adsp_common.a  -no-pie  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,--fatal-warnings && cd /__w/zephyr/zephyr/twister-out/intel_adsp_cavs15/samples/subsys/zbus/hello_world/sample.zbus.hello_world/zephyr && /usr/local/bin/cmake -E echo
/opt/toolchains/zephyr-sdk-0.14.2/xtensa-intel_apl_adsp_zephyr-elf/bin/../lib/gcc/xtensa-intel_apl_adsp_zephyr-elf/10.3.0/../../../../xtensa-intel_apl_adsp_zephyr-elf/bin/ld.bfd: warning: orphan section `._zbus_channel.static.acc_data' from `app/libapp.a(channels.c.obj)' being placed in section `._zbus_channel.static.acc_data'
/opt/toolchains/zephyr-sdk-0.14.2/xtensa-intel_apl_adsp_zephyr-elf/bin/../lib/gcc/xtensa-intel_apl_adsp_zephyr-elf/10.3.0/../../../../xtensa-intel_apl_adsp_zephyr-elf/bin/ld.bfd: warning: orphan section `._zbus_channel.static.version' from `app/libapp.a(channels.c.obj)' being placed in section `._zbus_channel.static.version'
/opt/toolchains/zephyr-sdk-0.14.2/xtensa-intel_apl_adsp_zephyr-elf/bin/../lib/gcc/xtensa-intel_apl_adsp_zephyr-elf/10.3.0/../../../../xtensa-intel_apl_adsp_zephyr-elf/bin/ld.bfd: warning: orphan section `._zbus_observer.static.my_subscriber' from `app/libapp.a(main.c.obj)' being placed in section `._zbus_observer.static.my_subscriber'
/opt/toolchains/zephyr-sdk-0.14.2/xtensa-intel_apl_adsp_zephyr-elf/bin/../lib/gcc/xtensa-intel_apl_adsp_zephyr-elf/10.3.0/../../../../xtensa-intel_apl_adsp_zephyr-elf/bin/ld.bfd: warning: orphan section `._zbus_observer.static.my_listener' from `app/libapp.a(main.c.obj)' being placed in section `._zbus_observer.static.my_listener'
/opt/toolchains/zephyr-sdk-0.14.2/xtensa-intel_apl_adsp_zephyr-elf/bin/../lib/gcc/xtensa-intel_apl_adsp_zephyr-elf/10.3.0/../../../../xtensa-intel_apl_adsp_zephyr-elf/bin/ld.bfd: zephyr/subsys/zbus/libsubsys__zbus.a(zbus.c.obj):(.literal.zbus_iterate_over_channels+0x0): undefined reference to `_zbus_channel_list_start'
/opt/toolchains/zephyr-sdk-0.14.2/xtensa-intel_apl_adsp_zephyr-elf/bin/../lib/gcc/xtensa-intel_apl_adsp_zephyr-elf/10.3.0/../../../../xtensa-intel_apl_adsp_zephyr-elf/bin/ld.bfd: zephyr/subsys/zbus/libsubsys__zbus.a(zbus.c.obj):(.literal.zbus_iterate_over_channels+0x4): undefined reference to `_zbus_channel_list_end'
/opt/toolchains/zephyr-sdk-0.14.2/xtensa-intel_apl_adsp_zephyr-elf/bin/../lib/gcc/xtensa-intel_apl_adsp_zephyr-elf/10.3.0/../../../../xtensa-intel_apl_adsp_zephyr-elf/bin/ld.bfd: zephyr/subsys/zbus/libsubsys__zbus.a(zbus.c.obj):(.literal.zbus_iterate_over_observers+0x0): undefined reference to `_zbus_observer_list_start'
/opt/toolchains/zephyr-sdk-0.14.2/xtensa-intel_apl_adsp_zephyr-elf/bin/../lib/gcc/xtensa-intel_apl_adsp_zephyr-elf/10.3.0/../../../../xtensa-intel_apl_adsp_zephyr-elf/bin/ld.bfd: zephyr/subsys/zbus/libsubsys__zbus.a(zbus.c.obj):(.literal.zbus_iterate_over_observers+0x4): undefined reference to `_zbus_observer_list_end'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

The documentation is not ready yet and needs a lot of revision. I am working on that this week. 

The code I guess it is in a stable version. Please take a look at that.

cfriedt reviewed (dismissed):
Looks great. I'm ok with this as is and will re-ack in case anyone insists on another rev to address compliance warnings, or abbreviate tests. Overall, great handling of this topic. Thanks for the work you've put into this @rodrigopex!

rodrigopex committed:
tests: zbus: Add zbus runtime observer registration tests

Add zbus subsystem test to verify the channel's runtime observers list
usage. It demonstrates the use of dynamically add and remove observer to
and from channels at runtime.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

carlocaione commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,253 @@
+
+#define _ZBUS_ASSERT_UNREACHABLE __ASSERT(0, "Reached unreachable code")
+
+#if defined(CONFIG_THREAD_NAME) && CONFIG_THREAD_NAME == 1
```
What is `CONFIG_THREAD_NAME == 1` for?

carlocaione commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,253 @@
+
+void __weak _zbus_notify_runtime_listeners(struct zbus_channel *chan)
+{
+}
+int __weak _zbus_notify_runtime_subscribers(struct zbus_channel *chan)
```
nit: add empty line

carlocaione commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,253 @@
+int __weak _zbus_notify_runtime_subscribers(struct zbus_channel *chan)
+{
+	return 0;
+}
+void _zbus_notify_listeners(struct zbus_channel *chan)
```
ditto

carlocaione commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,253 @@
+void _zbus_notify_listeners(struct zbus_channel *chan)
+{
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled) {
+			if ((*obs)->callback != NULL) {
```
just `&&` the two `if`s

carlocaione commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,253 @@
+	int err = 0;
+
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled) {
+			if ((*obs)->queue != NULL) {
```
ditto.

carlocaione commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,253 @@
+			}
+		}
+	}
+	err = _zbus_notify_runtime_subscribers(chan);
+	return err;
```
`return _zbus ...`

carlocaione commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,253 @@
+	}
+
+	err = k_mutex_lock(chan->mutex, timeout);
+	if (err) {
+		LOG_ERR("Could not pub, channel %s is busy.", chan->name);
+		_ZBUS_ASSERT_UNREACHABLE;
+		return err;
```
unlock the mutex on the error branch

carlocaione commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,253 @@
+		return err;
+	}
+
+	err = k_mutex_lock(chan->mutex, timeout);
+	if (err) {
+		LOG_ERR("Could not read, channel %s is busy.", chan->name);
+		_ZBUS_ASSERT_UNREACHABLE;
+		return err;
```
ditto

carlocaione commented:
Some nits but nothing really blocking

rodrigopex commented:
@carlocaione 
> Some nits but nothing really blocking

The last push resolved all the comments.

rodrigopex committed:
samples: zbus: Add runtime observer registration sample

The sample illustrates a way of using the runtime observer registration
feature. With this sample the developer can understand how to use static
and runtime observer registration.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

pdgendt commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,249 @@
+static inline int zbus_chan_pub_args_check(struct zbus_channel *chan, void *msg,
+					   k_timeout_t timeout)
+{
+	if (k_is_in_isr()) {
```
Could it be allowed to use `zbus` in `ISR` if `timeout == K_NO_WAIT`?

pdgendt commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,249 @@
+
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled && ((*obs)->queue != NULL)) {
+			err = k_msgq_put((*obs)->queue, &chan,
+					 K_MSEC(CONFIG_ZBUS_EVENT_DISPATCHER_QUEUE_TIMEOUT));
+			if (err) {
+				return err;
+			}
```
A single timeout option for all channels forces users to always take the worst case scenario into account. Maybe turn this into a channel member or pass it as a parameter when publishing.

Also not sure about the early return, if a single subscriber wasn't ready to receive the message, shouldn't all the others at least get it?

pdgendt commented on subsys/zbus/zbus_runtime_observers.c:
```diff
@@ -0,0 +1,120 @@
+#include <zephyr/zbus/zbus.h>
+LOG_MODULE_DECLARE(zbus, CONFIG_ZBUS_LOG_LEVEL);
+
+struct zbus_observer_node {
+	sys_snode_t node;
+	struct zbus_observer *obs;
+};
+
+static struct zbus_observer_node
+	_zbus_runtime_obs_list[CONFIG_ZBUS_RUNTIME_OBSERVERS_FREELIST_SIZE];
```
What is the rationale for keeping this in a separate buffer? Why not conditionally add the `sys_node_t node` to the `struct zbus_observer` if `CONFIG_ZBUS_RUNTIME_OBSERVERS_REGISTRATION` is set (like you do for the channel)?

pdgendt commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,411 @@
+ * @param chan The channel's reference.
+ * @param msg Reference to the message where the publish function copies the channel's
+ * message data from.
+ * @param timeout Waiting period to publish the channel,
+ *                or one of the special values K_NO_WAIT and K_FOREVER.
```
The `timeout` parameter is a bit misleading. It's currently used to lock the channel's mutex, but doesn't take the observer notify timeouts into account. The total call time can be greater than the timeout and still be successful.

rodrigopex commented:
@pdgendt thank you for your comments. Soon I will push changes related to the timing.

pdgendt commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,251 @@
+		if ((*obs)->enabled && ((*obs)->queue != NULL)) {
+			err = k_msgq_put(
+				(*obs)->queue, &chan,
+				K_TICKS((k_ticks_t)MAX(end_ticks - sys_clock_tick_get(), 0)));
```
For readability, please add a helper function, for example:
```c
static inline k_timeout_t timeout_remainder(uint64_t end)
{
	return K_TICKS((k_ticks_t)MAX(end - sys_clock_tick_get(), 0));
}
```

```suggestion
				timeout_remainder(end_ticks);
```

pdgendt commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,254 @@
+
+inline k_timeout_t _zbus_timeout_remainder(uint64_t end_ticks)
+{
+	return K_TICKS((k_ticks_t)MAX(end_ticks - sys_clock_tick_get(), 0));
```
Don't use functions as `MAX` macro arguments as these are going to be called twice.

```c
#define MAX(a, b) (((a) > (b)) ? (a) : (b))
```

pdgendt commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,411 @@
+{
+	int err = zbus_chan_pub_args_check(chan, msg, msg_size, timeout);
+	if (err) {
+		_ZBUS_ASSERT_UNREACHABLE;
```
I agree with @gmarull this shouldn't be unreachable.

pdgendt commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,254 @@
+	}
+	LOG_DBG("Channel %s published by thread %s %p", chan->name, _ZBUS_CURRENT_THREAD_INFO);
+	return err;
+}
```
I think the whole API can be simplified and use more zephyr provided utilities. Trimming `zbus_chan_pub` could come to something like:

```c
int zbus_chan_pub(struct zbus_channel *chan, void *msg, k_timeout_t timeout)
{
	int ret;
	uint64_t end_ticks = sys_clock_timeout_end_calc(timeout);

	__ASSERT(chan != NULL, "chan is required");
	__ASSERT(msg != NULL, "msg is required");
	__ASSERT(!k_is_in_isr())), "ISR blocked");

	if (chan->read_only) {
		return -EPERM;
	}

	if (chan->validator && !chan->validator(msg, chan->message_size)) {
		return -EACCESS;
	}

	ret = k_mutex_lock(chan->mutex, timeout);
	if (ret) {
		return ret;
	}

	memcpy(chan->message, msg, chan->message_size);

	_zbus_notify_listeners(chan);
	ret = _zbus_notify_subscribers(chan, end_ticks);

	k_mutex_unlock(chan->mutex);

	return ret;
}
```

pdgendt replied:
Checking for `msg == NULL` is arguable, if you would add a runtime check or `__ASSERT` as this can be passed along throughout the application code.

pdgendt commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,418 @@
+
+/** @cond INTERNAL_HIDDEN */
+
+#define _ZBUS_ASSERT_UNREACHABLE __ASSERT(0, "Reached unreachable code")
```
This is no longer used.

pdgendt commented on subsys/zbus/zbus_runtime_observers.c:
```diff
@@ -0,0 +1,144 @@
+};
+
+static struct zbus_observer_node
+	_zbus_runtime_obs_list[CONFIG_ZBUS_RUNTIME_OBSERVERS_FREELIST_SIZE];
```
I still don't get why this has to be a static list, you can add a `sys_node_t` member to `struct zbus_observer` based on the `Kconfig` option.

pdgendt replied:
Ping @rodrigopex

pdgendt commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,170 @@
+	int err;
+	uint64_t end_ticks = sys_clock_timeout_end_calc(timeout);
+
+	__ASSERT(!k_is_in_isr(), "ISR blocked");
```
I know this was in my example, but is there a reason why we it can't be allowed to run in interrupts if `K_TIMEOUT_EQ(timeout, K_NO_WAIT)`?

yperess reviewed (commented):
LGTM in regards to mocking the assert. I will also work on the side on having the system I described which will be more generic for other subsystems/drivers/etc to use.

yperess commented:
@rodrigopex I just realized that you can also use `CONFIG_ASSERT_TEST=y` to do this.

rodrigopex commented:
@yperess 

> @rodrigopex I just realized that you can also use `CONFIG_ASSERT_TEST=y` to do this.

I was trying to figure out how would I implement that using `CONFIG_ASSERT_TEST=y` but I could not think in a good way without interfering in other parts of the system. I want to suppress only the asserts inside the zbus code. I guess the current code is an acceptable solution. What do you think?

yperess commented:
> @yperess
> 
> > @rodrigopex I just realized that you can also use `CONFIG_ASSERT_TEST=y` to do this.
> 
> I was trying to figure out how would I implement that using `CONFIG_ASSERT_TEST=y` but I could not think in a good way without interfering in other parts of the system. I want to suppress only the asserts inside the zbus code. I guess the current code is an acceptable solution. What do you think?

Yes, what you have is perfectly acceptable. I just wanted to make sure you were aware of the config. You could still do as you wanted with CONFIG_ASSERT_TEST, but it would probably look odd (you'd basically need to check the file name in the `assert_post_action` and call `kpanic` if it was outside of your subsystem. Regardless, what you have works well for what you wanted to test :+1:

yperess commented:
@alevkoy I think you might be able to use this for your vbus design

pdgendt commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,170 @@
+		return -ENOMSG;
+	}
+
+	err = k_mutex_lock(chan->mutex, _zbus_timeout_remainder(end_ticks));
```
The first blocking call can actually use the `timeout` variable directly, no need to re-calculate here.

rodrigopex commented:
@yperess

> Yes, what you have is perfectly acceptable. I just wanted to make sure you were aware of the config. You could still do as you wanted with CONFIG_ASSERT_TEST, but it would probably look odd (you'd basically need to check the file name in the `assert_post_action` and call `kpanic` if it was outside of your subsystem. Regardless, what you have works well for what you wanted to test :+1: 

The issue with the current implementation is that it only does not blocks the rest of the execution. But it does tell nothing regarding the failure. I want to capture the failure in a zassert. Thanks for the comments. I will replace all zassert_true.

fabiobaltieri commented on doc/services/zbus/index.rst:
```diff
@@ -0,0 +1,356 @@
+    Zbus internals details.
+
+The bus makes the publish, read, and subscribe actions available over channels. Publishing and reading are available in all RTOS contexts except inside an Interrupt Service Routine (ISR). The publish and read operations were designed to be simple and fast; the procedure is a mutex locking followed by a memory copy to and from a shared memory region. Zbus observers' registration can be::
+* Static, defined in compile time. It is not possible to remove at runtime, but it is possible to suppress it by calling the :c:func:`zbus_obs_set_enable`;
```
This list is not rendering, I think it needs an empty space before.

fabiobaltieri commented on doc/services/zbus/index.rst:
```diff
@@ -0,0 +1,356 @@
+
+Zbus operation depends on channels and observers. Therefore, it is necessary to determine its message and observers list during the channel definition. A message is a regular C struct; the observer can be a subscriber (asynchronous) or a listener (synchronous). Channels can have a combination of the following characteristics:
+
+* ``Validator function`` makes a channel accepts only valid messages;
```
Not quite for the comment but: haven't seen any samples that tests the validator, would it make sense to add a validator function to one of those?

fabiobaltieri commented on doc/services/zbus/index.rst:
```diff
@@ -0,0 +1,356 @@
+                    zbus_chan_read(chan, &acc, K_NO_WAIT);
+                    LOG_DBG("From listener -> Acc x=%d, y=%d, z=%d", acc.x, acc.y, acc.z);
+            }
+    }
```
Should there be a `ZBUS_LISTENER_DEFINE(my_listener, listener_callback_example);` here?

fabiobaltieri commented on doc/services/zbus/index.rst:
```diff
@@ -0,0 +1,356 @@
+
+    bool control_validator(void* msg, size_t msg_size) {
+        struct control_msg* cm = msg;
+        bool is_valid = (msg->mode == -1) || (msg->mode == 0) || (msg->mode == 1);
```
`msg->move`?

fabiobaltieri commented on doc/services/zbus/index.rst:
```diff
@@ -0,0 +1,356 @@
+
+
+.. note::
+    Zbus can only use a limited number of dynamic observers. This number is 8 by default and can be changed by setting the value of :kconfig:option:`CONFIG_ZBUS_RUNTIME_OBSERVERS_FREELIST_SIZE` as needed. Use the :c:func:`zbus_obs_freelist_count` to verify how many available runtime observers slot exists.
```
I would avoid writing down the default config value here (i.e. drop the "This number is 8 by default". It'd be easy to forget to update this if we ever change the default, and the actual value is linked from the kconfig option name anyway.

fabiobaltieri commented on doc/services/zbus/index.rst:
```diff
@@ -0,0 +1,356 @@
+    {
+            struct zbus_channel *chan = NULL;
+
+            while (!k_msgq_get(my_subscriber.queue, &chan, K_FOREVER)) {
```
Have you considered adding a `zbus_...` function wrapping `k_msgq_get`? I'm going through the examples and it feels a bit out of place to use a msgq function to an internal field of `struct zbus_observer`, I'm thinking it may make sense to leave those fields opaque and have something that operates on the observer struct directly so you don't have to expose the `.queue` bit.

fabiobaltieri commented on tests/subsys/zbus/user_data/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+# SPDX-License-Identifier: Apache-2.0
+cmake_minimum_required(VERSION 3.13.1)
```
All other samples are using 3.20.0, can you adjust this and the others?

fabiobaltieri commented on tests/subsys/zbus/user_data/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+        "${CMAKE_CURRENT_SOURCE_DIR}/src/main.c"
+)
+
+include_directories(${HEADERS})
```
How would you feel about moving those local headers file in the same directory as the code (`src`) and dropping the include directory from the samples? Keep things a bit simpler, no other sample use a dedicated local include and personally for small samples like this I find it easier to have everything together.

fabiobaltieri commented on tests/subsys/zbus/user_data/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+)
+
+include_directories(${HEADERS})
+target_sources(app PRIVATE ${SOURCES})
```
Again I would just use

```
FILE(GLOB app_sources src/*.c)
target_sources(app PRIVATE ${app_sources})
```

Keep things a bit simpler.

fabiobaltieri commented on samples/subsys/zbus/remote_mock/boards/hifive1_revb.overlay:
```diff
@@ -0,0 +1,9 @@
+
+&uart1 {
+	status = "okay";
+    current-speed = <115200>;
+    clock-frequency = <16000000>;
```
Use tabs for these two.

fabiobaltieri commented on samples/subsys/zbus/remote_mock/remote_mock.py:
```diff
@@ -0,0 +1,51 @@
+# Copyright (c) 2022 Rodrigo Peixoto <rodrigopex@gmail.com>
```
Can you add a `#!/usr/bin/env python3` and `chmod +x` this?

(same for `uart_bridge/decoder.py`)

fabiobaltieri commented on samples/subsys/zbus/dyn_channel/src/consumer.c:
```diff
@@ -0,0 +1,36 @@
+
+#include <zephyr/logging/log.h>
+#include <zephyr/zbus/zbus.h>
+LOG_MODULE_DECLARE(zbus, CONFIG_ZBUS_LOG_LEVEL);
```
Empty line after this? It's a bit nitpicky but I would suggest going over the various file and using a coherent style for `LOG_MODULE_DECLARE`. The one I like is having `LOG_MODULE_DECLARE` together with the "include" and then an empty line and the rest, so that the ZBUS ones stand out more... but just pick a style and stick to it.

fabiobaltieri commented on samples/subsys/zbus/dyn_channel/src/consumer.c:
```diff
@@ -0,0 +1,36 @@
+	struct zbus_channel *chan = NULL;
+
+	while (!zbus_sub_wait(&consumer, &chan, K_FOREVER)) {
+		printk("Channel %s\n", chan->name);
```
Should this be a `LOG_INF`? (so it does not mix-up with the other LOG_)

fabiobaltieri commented on samples/subsys/zbus/work_queue/src/main.c:
```diff
@@ -0,0 +1,177 @@
+	struct sensor_msg msg = {0};
+
+	zbus_chan_read(sens->chan, &msg, K_MSEC(200));
+	printk("Sensor msg processed by WORK QUEUE handler dh%u: temp = %u, press = %u, "
```
`LOG_INF` instead of `printk`?

fabiobaltieri commented on samples/subsys/zbus/dyn_channel/src/main.c:
```diff
@@ -0,0 +1,73 @@
+
+void main(void)
+{
+	struct version_msg v = {0};
```
Drop the initialization if it's not needed (it gets immediately rewritten by `zbus_chan_read`)

fabiobaltieri commented on samples/subsys/zbus/dyn_channel/src/main.c:
```diff
@@ -0,0 +1,73 @@
+
+void filter_cb(struct zbus_channel *chan)
+{
+	struct external_data_msg *chan_message = NULL;
```
Initialization is not needed? You overwrite it just two lines below.

fabiobaltieri replied:
Still there :-)

fabiobaltieri commented on samples/subsys/zbus/dyn_channel/src/main.c:
```diff
@@ -0,0 +1,73 @@
+	struct external_data_msg *chan_message = NULL;
+
+	__ASSERT_NO_MSG(chan == &pkt_channel);
+	chan_message = (struct external_data_msg *)chan->message;
```
Is the casting needed? Think you can just use `chan_message = chan->message` since message is a `void *`.

fabiobaltieri commented on samples/subsys/zbus/dyn_channel/src/producer.c:
```diff
@@ -0,0 +1,30 @@
+void producer_thread(void)
+{
+	struct zbus_channel *chan = NULL;
+	void *msg = NULL;
```
Again, you should be able to drop these initializations.

fabiobaltieri commented on samples/subsys/zbus/remote_mock/src/mock_proxy.c:
```diff
@@ -0,0 +1,138 @@
+				 * other data at the buffer.
+				 */
+			} else {
+				if ('$' == net_buf_simple_pull_u8(rx_buf)) {
```
This is starting to get harsh with the indentation, how about moving this block in a separate function? (either from here or from line 104)

fabiobaltieri reviewed (commented):
Hey, did a pass on the sample code, added few comments. Some of those applies to other samples as well but I ran out of time so I'm sending what I have, spotted few others unnecessary automatic variable initializations that I think should be dropped. Could you make a pass on all sample files if you agree? It would also be nice to have coherent style for stuff like CMake files, prj.conf option grouping, headers etc... bit nitpicky but I think it helps when reading the various samples code.

Thanks for the great job so far!

rodrigopex commented:
> Hey, did a pass on the sample code, added few comments. Some of those applies to other samples as well but I ran out of time so I'm sending what I have, spotted few others unnecessary automatic variable initializations that I think should be dropped. Could you make a pass on all sample files if you agree? It would also be nice to have coherent style for stuff like CMake files, prj.conf option grouping, headers etc... bit nitpicky but I think it helps when reading the various samples code.

I will take a look and revisit all of them. Thanks for the comments.

rodrigopex commented:
> Hey, did a pass on the sample code, added few comments. Some of those applies to other samples as well but I ran out of time so I'm sending what I have, spotted few others unnecessary automatic variable initializations that I think should be dropped. Could you make a pass on all sample files if you agree?

@fabiobaltieri I tried to solve the points you said in all samples and tests as well. I have added some auxiliary functions to access channels and observers' struct fields like `zbus_cha_msg`, `zbus_chan_user_data`, `zbus_obs_name`, etc.

rodrigopex commented:
After installing the SDK version 0.15.0, I cannot generate the documentation or coverage info. Now, it is hard to check the code before pushing it.

rodrigopex committed:
samples: zbus: Add benchmark sample

The sample measures the time to transfer 256KB from the
producer to the consumers.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex commented:
Added benchmark sample.

fabiobaltieri commented on samples/subsys/zbus/hello_world/src/main.c:
```diff
@@ -0,0 +1,158 @@
+		 ZBUS_MSG_INIT(.x = 0, .y = 0, .z = 0) /* Initial value */
+);
+
+bool simple_chan_validator(void *msg, size_t msg_size)
```
This and most of the other sample code functions should be `static`.

fabiobaltieri commented on samples/subsys/zbus/hello_world/src/main.c:
```diff
@@ -0,0 +1,158 @@
+
+void subscriber_task(void)
+{
+	struct zbus_channel *chan = NULL;
```
More redundant initialization? (the line below updates it immediately right?)

fabiobaltieri commented on samples/subsys/zbus/hello_world/src/main.c:
```diff
@@ -0,0 +1,158 @@
+bool print_channel_data_iterator(struct zbus_channel *chan)
+{
+	LOG_INF("%d - Channel %s:", count, chan->name);
+	LOG_INF("      Message size: %d", chan->message_size);
```
Can you use `zbus_chan_name` and `zbus_chan_msg_size` for these? (thanks for adding those, good idea!)

fabiobaltieri commented on samples/subsys/zbus/hello_world/src/main.c:
```diff
@@ -0,0 +1,158 @@
+	LOG_INF("Sensor sample started raw reading, version %u.%u-%u!", v->major, v->minor,
+		v->build);
+
+#if defined(CONFIG_ZBUS_STRUCTS_ITERABLE_ACCESS) && CONFIG_ZBUS_STRUCTS_ITERABLE_ACCESS == 1
```
I think this should just be `#ifdef CONFIG_ZBUS_STRUCTS_ITERABLE_ACCESS`. It's a bool so it should be either defined or not.

fabiobaltieri commented on samples/subsys/zbus/hello_world/src/main.c:
```diff
@@ -0,0 +1,158 @@
+
+bool print_observer_data_iterator(struct zbus_observer *obs)
+{
+	LOG_INF("%d - %s %s", count, ((obs->queue != NULL) ? "Subscriber" : "Listener"), obs->name);
```
Drop the parenthesis from the ternary? (not sure it's a guideline but, you are putting double parenthesis here and no parenthesis on the one of 156) Personaly I would drop the `!= NULL` as well.

`LOG_INF("%d - %s %s", count, obs->queue ? "Subscriber" : "Listener", obs->name);` <- should work fine and is more compact

fabiobaltieri commented on samples/subsys/zbus/hello_world/src/main.c:
```diff
@@ -0,0 +1,158 @@
+	err = zbus_chan_pub(&simple_chan, &value, K_MSEC(200));
+
+	LOG_INF("Pub an invalid value to a channel with validator. Return %serror",
+		err == -ENOMSG ? "" : "no ");
```
I find this pattern of capturing the error code and reducing it to a "yes/no" a bit awkard. Imagine this failing with another error code and having to debug it, you'd end up having to change the sample to print the actual code. I would rather just log the return value as it is, (like instead of `Return %serror` something like `ret = %d` and the code). Easier to spot than a "no" in the middle of the log message as well. Same for the one above.

rodrigopex replied:
Fixed

fabiobaltieri commented on samples/subsys/zbus/work_queue/src/main.c:
```diff
@@ -0,0 +1,183 @@
+
+void thread_handler1_task(void)
+{
+	struct zbus_channel *chan = NULL;
```
Unnecessary initialization (below as well)

fabiobaltieri commented on samples/subsys/zbus/work_queue/src/main.c:
```diff
@@ -0,0 +1,183 @@
+	struct zbus_channel *chan = NULL;
+
+	while (!zbus_sub_wait(&thread_handler1_sub, &chan, K_FOREVER)) {
+		struct sensor_msg msg = {0};
```
same

fabiobaltieri commented on samples/subsys/zbus/dyn_channel/src/consumer.c:
```diff
@@ -0,0 +1,44 @@
+	struct zbus_channel *chan;
+
+	while (!zbus_sub_wait(&consumer_sub, &chan, K_FOREVER)) {
+		LOG_INF("Channel %s\n", chan->name);
```
`zbus_chan_name`?

fabiobaltieri commented on samples/subsys/zbus/dyn_channel/src/consumer.c:
```diff
@@ -0,0 +1,44 @@
+		if (!zbus_chan_claim(&pkt_chan, K_MSEC(100))) {
+			chan_message = zbus_chan_msg(&pkt_chan);
+
+			LOG_WRN("size=%02d", (int)chan_message->size);
```
`zbus_chan_msg_size`?

fabiobaltieri commented on samples/subsys/zbus/uart_bridge/src/core.c:
```diff
@@ -0,0 +1,37 @@
+	struct action_msg start = {false};
+
+	while (1) {
+		LOG_DBG("Core sending start measurement with status %d", start.status);
```
Is this correct? You say sending with status.status but then you invert it and send, maybe you should send and then invert?

fabiobaltieri commented on samples/subsys/zbus/remote_mock/src/mock_proxy.c:
```diff
@@ -0,0 +1,148 @@
+
+void proxy_callback(struct zbus_channel *chan)
+{
+	bool *generated_by_the_bridge = (bool *)zbus_chan_user_data(chan);
```
Drop the casting? Should be redundant.

fabiobaltieri commented on samples/subsys/zbus/uart_bridge/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -4,5 +4,5 @@
 &uart1 {
 	status = "okay";
-    current-speed = <115200>;
+	current-speed = <115200>;
```
This should have gone to the previous patch.

fabiobaltieri replied:
still applicable :-)

fabiobaltieri commented on samples/subsys/zbus/benchmark/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+project(benchmark)
+
+zephyr_compile_definitions(
+        -DBM_MESSAGE_SIZE=${MSG_SIZE}U -DBM_ONE_TO=${ONE_TO}LLU -DBM_ASYNC=${ASYNC}U
```
These should really be sample specific Kconfig options (`ZBUS_SAMPLE_BENCHMARK_MSG_SIZE` etc...). Then you won't have to use `extra_args` but you can use `extra_configs` instead.

fabiobaltieri commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -0,0 +1,216 @@
+	{                                                                                          \
+		struct external_data_msg *actual_message_data;                                     \
+		struct zbus_channel *chan;                                                         \
+		struct bm_msg msg_received = {0};                                                  \
```
drop the init

fabiobaltieri commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -0,0 +1,216 @@
+#endif
+#endif
+
+struct bm_msg msg_received = {0};
```
drop the (which in this case is double redundant, globals are all zeroed at startup), also `static` (with most functions here as well I guess, see my comment on the first sample)

fabiobaltieri commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -0,0 +1,216 @@
+		BM_ONE_TO, BM_ASYNC ? "A" : "", BM_MESSAGE_SIZE);
+
+	struct bm_msg msg = {0};
+	struct external_data_msg *actual_message_data = NULL;
```
Again drop the init, it's assigned just few lines below

fabiobaltieri reviewed (commented):
Thanks for the followup, few more comments on the samples, more or less nitpicky. :-)

Would hold on adding more code for now, it's a big enough PR to review as it is.

rodrigopex commented:
@fabiobaltieri 

> Thanks for the followup, few more comments on the samples, more or less nitpicky. :-)
> 
> Would hold on adding more code for now, it's a big enough PR to review as it is.

No more code, I promess :)

fabiobaltieri commented on doc/services/zbus/index.rst:
```diff
@@ -0,0 +1,378 @@
+                    LOG_DBG("From listener -> Acc x=%d, y=%d, z=%d", acc->x, acc->y, acc->z);
+            }
+    }
+User Data
```
Doc build seems to want a blank line here:

```
/home/runner/work/zephyr/zephyr/doc/_build/src/services/zbus/index.rst:300:Explicit markup ends without a blank line; unexpected unindent.
```
```

rodrigopex commented:
Hi, community. My baby son was born last Thursday. I will get back to work on this PR as soon as possible (next week, I guess). For now, I am taking care of my wife and son. Thanks.

rodrigopex commented:
Got back to work! :muscle:

fabiobaltieri commented on samples/subsys/zbus/work_queue/src/main.c:
```diff
@@ -0,0 +1,183 @@
+		 ZBUS_MSG_INIT(0)		      /* Initial value {0} */
+);
+
+void fh1_cb(struct zbus_channel *chan)
```
These should be all static (listener handles, thread handlers etc). Same as in other samples.

fabiobaltieri commented on samples/subsys/zbus/dyn_channel/src/producer.c:
```diff
@@ -0,0 +1,36 @@
+	void *msg;
+	int i = 0;
+
+	do {
```
Should this be a for loop? With a separate exit path for the `!zbus_sub_wait` where you could put an explicit log as well.

fabiobaltieri commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -0,0 +1,216 @@
+		 NULL, /* Validator */
+		 ZBUS_OBSERVERS(s1
+
+#if (CONFIG_BM_ONE_TO >= 2LLU)
+				,
+				s2
+#if (CONFIG_BM_ONE_TO > 2LLU)
+				,
+				s3, s4
+#if (CONFIG_BM_ONE_TO > 4LLU)
+				,
+				s5, s6, s7, s8
+#if (CONFIG_BM_ONE_TO > 8LLU)
+				,
+				s9, s10, s11, s12, s13, s14, s15, s16
+#endif
+#endif
+#endif
+#endif
```
Something for a followup - let's not add more feature to this - it would be cool to have a way of defining observers throughtout the code with iterable sections instead of an explicit references. That is, a set of macros to declare the memory section with list of observer structures, and then something to define the observer so that it ends up there an gets picked up for the channel without an explicit reference. That would clean up patterns like this, and would allow spreading the subscribers/listeners around multiple source file and enabling/disabling them statically by just excluding the specific source from the build, as it's done with most ohter low level structures in Zephyr.

Again - just something to discuss as a followup.

fabiobaltieri commented on tests/subsys/zbus/unittests/src/main.c:
```diff
@@ -0,0 +1,598 @@
+#include <zephyr/ztest.h>
+LOG_MODULE_DECLARE(zbus, CONFIG_ZBUS_LOG_LEVEL);
+
+bool hard_msg_validator(void *msg, size_t msg_size)
```
All the local test functions and variables should be `static` as well.

fabiobaltieri commented on tests/subsys/zbus/dyn_channel/src/main.c:
```diff
@@ -0,0 +1,144 @@
+static struct {
+	uint8_t a;
+	uint64_t b;
+} my_random_data_expected = {0};
```
Drop the initialization to zero, all global variables are zeroed automatically, saves few bytes of flash.

fabiobaltieri replied:
hey this are still here

fabiobaltieri commented on tests/subsys/zbus/unittests/src/main.c:
```diff
@@ -0,0 +1,598 @@
+
+static enum operation current_isr_operation = NONE;
+
+static struct action_msg ga = {0};
```
Drop the initialization.

fabiobaltieri replied:
Think you missed this one. :-)

fabiobaltieri commented on tests/subsys/zbus/unittests/src/main.c:
```diff
@@ -0,0 +1,598 @@
+	struct k_work work;
+};
+
+static struct aux2_wq_data wq_handler = {0};
```
Drop the initialization.

fabiobaltieri commented on tests/subsys/zbus/integration/src/main.c:
```diff
@@ -0,0 +1,303 @@
+K_THREAD_DEFINE(core_thread_id, 1024, core_thread, NULL, NULL, NULL, 3, 0, 0);
+
+static int count_net;
+static struct net_pkt_msg pkt = {0};
```
Drop the initialization.

fabiobaltieri commented on tests/subsys/zbus/runtime_observers_registration/src/main.c:
```diff
@@ -0,0 +1,158 @@
+	struct k_work work;
+};
+
+static struct aux2_wq_data wq_handler = {0};
```
Drop the initialization.

fabiobaltieri reviewed (commented):
Hey just spotted few unnecessary initialization to 0 for global variables, happy enough besides that, looks pretty polished to me. Good job!

rodrigopex commented:
> Hey just spotted few unnecessary initialization to 0 for global variables, happy enough besides that, looks pretty polished to me. Good job!

Thanks a lot for your contributions @fabiobaltieri!

MaureenHelm reviewed (dismissed):
> looks pretty polished to me. Good job!

Totally agree. Usually PRs of this size are painful to review, but this one reads very easily. Thank you!

> Hi, community. My baby son was born last Thursday. I will get back to work on this PR as soon as possible (next week, I guess). For now, I am taking care of my wife and son. Thanks.

Congratulations!

anangl commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,507 @@
+	static _type _CONCAT(_zbus_message_, _name) = _init_val;                                   \
+	K_MUTEX_DEFINE(_CONCAT(_zbus_mutex_, _name));                                              \
+	FOR_EACH_NONEMPTY_TERM(_ZBUS_OBS_EXTERN, (;), _observers)                                 \
+	_ZBUS_STRUCT_DECLARE(zbus_channel, _name) = {                                              \
```
Most fields of this structure is not supposed to be modified at runtime (I assume the only exception is `runtime_observers`; or am I wrong and the intention is to allow also modification of `user_data` or `read_only`?), why not try to make it `const` and locate in ROM?

anangl commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,507 @@
+ */
+#define ZBUS_CHAN_DEFINE(_name, _read_only, _type, _validator, _observers, _init_val)              \
+	static _type _CONCAT(_zbus_message_, _name) = _init_val;                                   \
+	K_MUTEX_DEFINE(_CONCAT(_zbus_mutex_, _name));                                              \
```
Why not `static`?

anangl replied:
Yes.

rodrigopex replied:
I did not use that because I would need to call the init function. Suggestions? @anangl

rodrigopex replied:
Done

anangl commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,210 @@
+	return 0;
+}
+
+void _zbus_notify_listeners(struct zbus_channel *chan)
+{
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled && ((*obs)->callback != NULL)) {
+			(*obs)->callback(chan);
```
Have you considered using separate arrays of listeners and subscribers? Currently, the observers array is always iterated twice on each publishing to the channel, even when there is no listeners or no subscribers.

rodrigopex replied:
For runtime observers, it will be necessary to separate as well to keep consistency. Do you think it would be interesting? We will need to have  `add_sub` and `add_lis` instead of `add_obs`. @anangl

anangl commented on subsys/zbus/Kconfig:
```diff
@@ -0,0 +1,57 @@
+	  and recycled after use. The technique avoids dynamic allocation and allows the code to increase the
+	  number of observers by only changing a configuration.
+
+config ZBUS_CHANNEL_USER_DATA
```
How about getting rid of this option and using `(CONFIG_ZBUS_CHANNEL_USER_DATA_SIZE > 0)` to check if the `user_data` field is present?

anangl commented on subsys/zbus/Kconfig:
```diff
@@ -0,0 +1,57 @@
+
+config ZBUS_RUNTIME_OBSERVERS_REGISTRATION
+	bool "Zbus runtime observer registration."
+	default n
```
`default n` is redundant.

anangl commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,210 @@
+	return 0;
+}
+
+void _zbus_notify_listeners(struct zbus_channel *chan)
```
All functions not supposed to be used externally should be `static`.

anangl replied:
That's the reason why it is not good to have them declared as `__weak`.

anangl commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,210 @@
+	return K_TICKS((k_ticks_t)MAX(end_ticks - now_ticks, 0));
+}
+
+void __weak _zbus_notify_runtime_listeners(struct zbus_channel *chan)
```
Do you need to use `__weak` for this? Wouldn't `#if` suffice?

anangl replied:
A lot? A few, I think.

rodrigopex replied:
Removed.

anangl commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,507 @@
+ */
+struct zbus_channel {
+	/** Channel name. */
+	const char *name;
```
Shouldn't this field be optional (in `zbus_observer` as well)? I guess it's only for debugging/logging purposes.

desowin commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,560 @@
+ * @retval 0 Observer added to the channel.
+ * @retval -EINVAL Some parameter is invalid.
+ */
+int zbus_chan_add_obs(struct zbus_channel *chan, struct zbus_observer *obs, k_timeout_t timeout);
```
Why use the observers pool? Why not make the runtime observer provide the memory for `struct zbus_observer_node`, for example in a macro like `ZBUS_RUNTIME_SUBSCRIBER_DEFINE()`?

rodrigopex replied:
@desowin ping

desowin replied:
Why would you want to add the same observer to multiple channels? Shouldn't there be a 1:1 relationship between observer and a channel?

rodrigopex replied:
Not at all. The relationship can be 1:1, 1:N, and N:N between publishers and observers. You can see the same observer added to several channels. It would be painful to have an observer for each channel we would like to observe. It would increase the code size unnecessarily.

desowin commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,411 @@
+
+K_MSGQ_DEFINE(_zbus_channels_changed_msgq, sizeof(enum zbus_chan_idx), 32, 2);
+
+#define _ZBUS_ASSERT_UNREACHABLE __ASSERT(0, "Reached unreachable code")
```
Why would you want to disable it? I see no reason at all to disable `CODE_UNREACHABLE` under any circumstances.

desowin commented on doc/services/zbus/index.rst:
```diff
@@ -0,0 +1,405 @@
+Iterating over channels and observers
+=====================================
+
+There is an iterator mechanism in zbus that enables the developer to execute some procedure per channel and observer. The sequence executed is sorted by channel or observer name.
```
But this only applies to the statically defined observers? The runtime ones would not be sorted, right?

rodrigopex replied:
No. Indeed I did not add the runtime observers to it. Do you think it might be good to do so? I can add that.

desowin replied:
When the documentation mentions "per channel and observer", I expect it means all observers and not only static ones.

rodrigopex replied:
You are right. I will fix that.

rodrigopex replied:
That is already fixed.

rodrigopex commented on doc/services/zbus/index.rst:
```diff
@@ -0,0 +1,405 @@
+            uint16_t build;
+    };
+
+    ZBUS_CHAN_DEFINE(version_chan,	       /* Name */
```
Ok. I will fix that.

rodrigopex replied:
Fixed

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,224 @@
+	}
+
+	if (chan->validator != NULL && !chan->validator(msg, chan->message_size)) {
+		return -ENOMSG;
```
So you prefer offensive to defensive programming, as do I. But in some scenarios, breaking the code would not be good, so the developer may have a way to adjust something and keep the code working. So I am trying to be as flexible as possible.

anangl commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,210 @@
+}
+
+#if defined(CONFIG_ZBUS_CHANNEL_USER_DATA)
+inline void *zbus_chan_user_data(struct zbus_channel *chan)
```
> Should I remove it, right?

No, rather provide such functions as `static inline` in the header file.

rodrigopex replied:
Fixed. Done.

anangl commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,507 @@
+ * @retval -EAGAIN Waiting period timed out.
+ * @retval -EINVAL Some parameter is invalid.
+ */
+int zbus_chan_pub(struct zbus_channel *chan, void *msg, k_timeout_t timeout);
```
> The channel cannot be const because it is being modified by the function.

`zbus_chan_pub`? What does it need to modify in the channel structure?

rodrigopex replied:
SYS_SLIST_FOR_EACH_CONTAINER_SAFE complains about the const modifier related to the static observers, and when enabled, the user_data is modified per channel.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,224 @@
+	memcpy(chan->message, msg, chan->message_size);
+
+	_zbus_notify_listeners(chan);
+	err = _zbus_notify_subscribers(chan, end_ticks);
```
I will remove the return at that point. An assert could catch that during the development process. What do you think?

rodrigopex replied:
I put them together. The logic is to return the last error but not stop the execution. The error return in this case, when the `CONFIG_ZBUS_ASSERT_MOCK` is enabled, is `-EFAULT`. However, during the execution, the LOG_ERR will be printed, and the last error code from the execution will be returned. I could not see a simple way of doing that. Now the notification does not stop at the first error. I cannot return zero (no error) if the notification is not complete.

rodrigopex replied:
Adjuted. Please take a look @anangl

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,224 @@
+		if ((*obs)->enabled && ((*obs)->queue != NULL)) {
+			err = k_msgq_put((*obs)->queue, &chan, _zbus_timeout_remainder(end_ticks));
+			if (err) {
+				return err;
```
Done. I have changed the way of the notification works. Please take a look at that. @anangl

rodrigopex commented on subsys/zbus/zbus_runtime_observers.c:
```diff
@@ -0,0 +1,153 @@
+		}
+	}
+
+	err = k_mutex_lock(chan->mutex, timeout);
```
Done. I have added mutex protection to the runtime observer pool.

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,507 @@
+	const char *name;
+
+	/** Read only flag. Indicates the channel cannot receive publications. */
+	bool read_only;
```
I will remove the read-only channels.

rodrigopex commented on doc/services/zbus/index.rst:
```diff
@@ -0,0 +1,405 @@
+
+    The ``Validator function`` will not take effect in read-only channels.
+
+The following code defines and initializes a read-only channel and its dependencies. This channel represents a firmware version, for example.
```
I will remove that.

rodrigopex replied:
Removed. Not read-only channels anymore.

rodrigopex reviewed (commented):
I commented on some points. Others I fixed.

rodrigopex reviewed (commented):
I fixed several points and commented other, please let me know what do you think.

rodrigopex reviewed (commented):
Done.

rodrigopex commented:
Major changes from the last push:

1. Removed support for read-only channels;
2. I did change a little the way of using `user_data`. I removed the data allocation from zbus code side and added an argument in the `ZBUS_CHANNEL_DEFINE` for adding the `user_data` pointer.

@anangl @desowin

rodrigopex reviewed (commented):
Now the runtime observers pool uses Zephyr memory slabs. This decreases complexity and avoids reinventing the wheel.

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,555 @@
+#define ZBUS_RUNTIME_OBSERVERS_LIST_INIT                                                           \
+	.runtime_observers = SYS_SLIST_STATIC_INIT(&_name.runtime_observers),
+#else
+#define ZBUS_RUNTIME_OBSERVERS_LIST_INIT /* No user data */
```
Not anymore. It would be better to leave the developers to allocate the `user data`. The `user data` field only points to it during the channel's definition macro. Doing this allows each channel to have independent `user data`, which was impossible before. The user data sample illustrates how to do that.

anangl replied:
I meant the comment. It does not seem to be related to the `ZBUS_RUNTIME_OBSERVERS_LIST_INIT` macro. It looks like some copy-paste remnant.

rodrigopex replied:
Indeed. I will remove that.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,177 @@
+}
+
+#if (CONFIG_ZBUS_RUNTIME_OBSERVERS_POOL_SIZE > 0)
+void _zbus_notify_runtime_listeners(struct zbus_channel *chan);
+
+int _zbus_notify_runtime_subscribers(struct zbus_channel *chan, uint64_t end_ticks);
```
I did that to separate the runtime observer logic from the actual code. By I can do that using ifdef as well. I will apply this change.

rodrigopex replied:
done

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,555 @@
+ * @param timeout Waiting period for a notification arrival,
+ *                or one of the special values K_NO_WAIT and K_FOREVER.
+ *
+ * @retval 0 Notification received.
+ * @retval -EBUSY The subscriber queue is busy.
+ * @retval -EAGAIN Waiting period timed out.
+ * @retval -EINVAL Some parameter is invalid.
+ */
+int zbus_sub_wait(struct zbus_observer *sub, struct zbus_channel **chan, k_timeout_t timeout);
```
Maybe it is outdated, it has changed after the first document writing. I will double-check that.

rodrigopex replied:
Fixed.

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,555 @@
+ * @retval -ENOMSG The message is invalid based on the validator function.
+ * @retval -EBUSY The channel is busy.
+ * @retval -EAGAIN Waiting period timed out.
+ * @retval -EFAULT Could not notify all the observers.
+ * @retval -EINVAL Some parameter is invalid.
+ */
+int zbus_chan_pub(struct zbus_channel *chan, const void *msg, k_timeout_t timeout);
```
I will double-check that.

rodrigopex replied:
Fixed to other functions as well.

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,555 @@
+	const char *name;
+#endif
+	/** Message size. Represents the channel's message size. */
+	const uint16_t message_size;
```
Great idea. I will do that.

rodrigopex replied:
@anangl How to use `const` structs with `STRUCT_SECTION_ITERABLE`? I need that to accomplish the suggestion.

anangl replied:
I think I don't understand what the problem is. Why do you need to add a `const` there? I think you can go with `STRUCT_SECTION_FOREACH(zbus_channel, chan)`, as it is now, and just update the `iterator_func` type. Wouldn't that work?

rodrigopex replied:
I could make the whole channel constant using that approach to separate the syslist data and point to that. `SYS_SLIST_FOR_EACH_CONTAINER_SAFE` is used in several functions.

rodrigopex replied:
Done.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,177 @@
+	return err;
+}
+
+int zbus_obs_set_enable(struct zbus_observer *obs, bool enabled)
+{
+	_ZBUS_ASSERT(obs != NULL, "obs is required");
+
+	obs->enabled = enabled;
+
+	return 0;
+}
```
Done.

rodrigopex reviewed (commented):
The zbus_channel struct is now const. The functions that receive a channel also have a const pointer argument. Some documentation related to `-EFAULT` were fixed as well.

rodrigopex commented:
@anangl @desowin Is there anything I can do? This PR is waiting for your change request acceptance. Thanks.

fabiobaltieri commented:
> @anangl @desowin Is there anything I can do? This PR is waiting for your change request acceptance. Thanks.

Can you check on the failing CI run? It breaks on a couple of zbus specific tests.

anangl commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,653 @@
+		.message = &_CONCAT(_zbus_message_, _name),  /* Reference to the message */        \
+		.validator = (_validator),		     /* Validator function */              \
+		.mutex = &_CONCAT(_zbus_mutex_, _name),	     /* Channel's Mutex */                 \
+		.observers = (const struct zbus_observer **)(const struct zbus_observer *[]){      \
+			FOR_EACH_NONEMPTY_TERM(ZBUS_REF, (,), _observers)                        \
+				NULL}} /* Static observer list */
```
Why not define above:
```
	static const struct zbus_observer * _CONCAT(_zbus_observers_, _name) = {		   \
		FOR_EACH_NONEMPTY_TERM(ZBUS_REF, (,), _observers)				   \
		NULL										   \
	};											   \
```
and here use just:
```
		.observers = &_CONCAT(_zbus_observers_, _name) /* Static observer list */          \
```

rodrigopex replied:
Added.

anangl commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,653 @@
+	ZBUS_RUNTIME_OBSERVERS_LIST_DECL(_CONCAT(_runtime_observers_, _name));                     \
+	FOR_EACH_NONEMPTY_TERM(_ZBUS_OBS_EXTERN, (;), _observers)                                 \
+	const _ZBUS_STRUCT_DECLARE(zbus_channel, _name) = {                                        \
+		ZBUS_CHANNEL_NAME_INIT(_name)	 /* Name */                                        \
+			.user_data = _user_data, /* User data */                                   \
+		ZBUS_RUNTIME_OBSERVERS_LIST_INIT(                                                  \
+			_CONCAT(_runtime_observers_, _name)) /* Runtime observer list */           \
+			.message_size = sizeof(_type),	     /* Message size */                    \
+		.message = &_CONCAT(_zbus_message_, _name),  /* Reference to the message */        \
```
Could you correct the alignment here and in other places where `_ZBUS_STRUCT_DECLARE()` is used??

rodrigopex replied:
This is a clang-format issue. I can do that, but every time I run the clang-format, it will make that mess again.

anangl replied:
Why use a tool that does not do its job well?

rodrigopex commented on subsys/zbus/zbus_runtime_observers.c:
```diff
@@ -0,0 +1,110 @@
+
+LOG_MODULE_DECLARE(zbus, CONFIG_ZBUS_LOG_LEVEL);
+
+K_MEM_SLAB_DEFINE(_zbus_runtime_obs_pool, sizeof(struct zbus_observer_node),
```
Fixed

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,653 @@
+ * @retval -EAGAIN Waiting period timed out.
+ * @retval -EINVAL Some parameter is invalid.
+ */
+int zbus_chan_add_obs(const struct zbus_channel *chan, struct zbus_observer *obs,
```
Added

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,653 @@
+	struct zbus_observer *obs;
+};
+
+static inline void _zbus_notify_runtime_listeners(const struct zbus_channel *chan)
```
You are right. Fixed.

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,653 @@
+
+struct zbus_observer_node {
+	sys_snode_t node;
+	struct zbus_observer *obs;
```
Fixed

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,169 @@
+	return K_TICKS((k_ticks_t)MAX(end_ticks - now_ticks, 0));
+}
+
+#if (CONFIG_ZBUS_RUNTIME_OBSERVERS_POOL_SIZE > 0)
+void _zbus_notify_runtime_listeners(const struct zbus_channel *chan);
+
+int _zbus_notify_runtime_subscribers(const struct zbus_channel *chan, uint64_t end_ticks);
+#endif /* CONFIG_ZBUS_RUNTIME_OBSERVERS_POOL_SIZE */
```
Fixed

rodrigopex commented on subsys/zbus/zbus_runtime_observers.c:
```diff
@@ -0,0 +1,110 @@
+	/* Check if the observer is already a runtime observer */
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(chan->runtime_observers, obs_nd, tmp, node) {
+
+		__ASSERT(obs_nd != NULL, "observer node is NULL");
```
Supposedly never. Removed.

rodrigopex commented on subsys/zbus/zbus_runtime_observers.c:
```diff
@@ -0,0 +1,110 @@
+	_ZBUS_ASSERT(chan != NULL, "chan is required");
+	_ZBUS_ASSERT(obs != NULL, "obs is required");
+
+	err = k_mutex_lock(chan->mutex, _zbus_timeout_remainder(end_ticks));
```
Fixed.

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,653 @@
+	/** Access control mutex. Points to the mutex used to avoid race conditions
+	 * for accessing the channel.
+	 */
+	struct k_mutex *mutex;
```
The observers field is already const. The other two cannot be const because of these errors:

```console
/.../zephyrproject/zephyr/subsys/zbus/zbus_runtime_observers.c:41:44: warning: passing argument 1 of ‘k_mutex_unlock’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   41 |                         k_mutex_unlock(chan->mutex);
      |                                        ~~~~^~~~~~~
```

```console
/home/rodrigopex/Cloud/Projects/zephyrproject/zephyr/subsys/zbus/zbus_runtime_observers.c:39:47: warning: passing argument 1 of ‘sys_slist_peek_head’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   39 |         SYS_SLIST_FOR_EACH_CONTAINER_SAFE(chan->runtime_observers, obs_nd, tmp, node) {
      |                                           ~~~~^~~~~~~~~~~~~~~~~~~
/.../zephyrproject/zephyr/include/zephyr/sys/util.h:211:29: note: in definition of macro ‘CONTAINER_OF’
  211 |         ((type *)(((char *)(ptr)) - offsetof(type, field)))
      |                             ^~~
/.../zephyrproject/zephyr/include/zephyr/sys/list_gen.h:35:9: note: in expansion of macro ‘Z_GENLIST_CONTAINER’
   35 |         Z_GENLIST_CONTAINER(sys_ ## __lname ## _peek_head(__l), __cn, __n)
      |         ^~~~~~~~~~~~~~~~~~~
/.../zephyrproject/zephyr/include/zephyr/sys/list_gen.h:52:21: note: in expansion of macro ‘Z_GENLIST_PEEK_HEAD_CONTAINER’
   52 |         for (__cn = Z_GENLIST_PEEK_HEAD_CONTAINER(__lname, __l, __cn, __n),   \
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/.../zephyrproject/zephyr/include/zephyr/sys/slist.h:181:9: note: in expansion of macro ‘Z_GENLIST_FOR_EACH_CONTAINER_SAFE’
  181 |         Z_GENLIST_FOR_EACH_CONTAINER_SAFE(slist, __sl, __cn, __cns, __n)
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/.../zephyrproject/zephyr/subsys/zbus/zbus_runtime_observers.c:39:9: note: in expansion of macro ‘SYS_SLIST_FOR_EACH_CONTAINER_SAFE’
   39 |         SYS_SLIST_FOR_EACH_CONTAINER_SAFE(chan->runtime_observers, obs_nd, tmp, node) {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /.../zephyrproject/zephyr/include/zephyr/kernel_includes.h:25,
                 from /.../zephyrproject/zephyr/include/zephyr/kernel.h:17,
                 from /.../zephyrproject/zephyr/subsys/zbus/zbus_runtime_observers.c:5:
/.../zephyrproject/zephyr/include/zephyr/sys/slist.h:231:61: note: expected ‘sys_slist_t *’ {aka ‘struct _slist *’} but argument is of type ‘const sys_slist_t *’ {aka ‘const struct _slist *’}
  231 | static inline sys_snode_t *sys_slist_peek_head(sys_slist_t *list)
      |                                                ~~~~~~~~~~~~~^~~~
```

anangl replied:
> The observers field is already const.

No, it's not. And also the array it points to is not, though it should be. Currently only the structures pointed to by items of the observers array are `const`. That's what you should use:
```
	const struct zbus_observer *const *const observers;
```

> The other two cannot be const because of these warnings:

You get the warnings because you are not making the fields `const` but the values they point to (which indeed should not be `const`, because they are supposed to be modified).
`const struct k_mutex *mutex;` has different meaning than `struct k_mutex *const mutex;`.

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,653 @@
+ *                or one of the special values K_NO_WAIT and K_FOREVER.
+ *
+ * @retval 0 Channel published.
+ * @retval -ENOMSG The message is invalid based on the validator function.
```
Fixed

rodrigopex commented:
> Can you check on the failing CI run? It breaks on a couple of zbus specific tests.

@fabiobaltieri, for some unknown reason, the same sample running on another RISCV boards is breaking at the `hifive_unleashed`. I have added that to the `platform_exclude` in the sample file.

The other CI issue is related to the `_ZBUS_ASSERT`, which has already been discussed before and is acceptable. I have submitted another revision. Let's see if it passes now.

anangl commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,615 @@
+		static const struct zbus_observer *_CONCAT(_zbus_observers_, _name)[] = {          \
+			FOR_EACH_NONEMPTY_TERM(ZBUS_REF, (,), _observers) NULL};                 \
+	const _ZBUS_STRUCT_DECLARE(zbus_channel, _name) = {                                        \
+		ZBUS_CHANNEL_NAME_INIT(_name)	 /* Name */                                        \
+			.user_data = _user_data, /* User data */                                   \
+		ZBUS_RUNTIME_OBSERVERS_LIST_INIT(                                                  \
+			_CONCAT(_runtime_observers_, _name))   /* Runtime observer list */         \
+			.message_size = sizeof(_type),	       /* Message size */                  \
+		.message = &_CONCAT(_zbus_message_, _name),    /* Reference to the message */      \
+		.validator = (_validator),		       /* Validator function */            \
+		.mutex = &_CONCAT(_zbus_mutex_, _name),	       /* Channel's Mutex */               \
+		.observers = _CONCAT(_zbus_observers_, _name)} /* Static observer list */
```
It would be nice to initialize the fields here in the order they are defined in the `zbus_channel` structure, so that this macro could be used also from C++ code.

rodrigopex replied:
I will fix that.

rodrigopex replied:
Done

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -0,0 +1,615 @@
+	static K_MUTEX_DEFINE(_CONCAT(_zbus_mutex_, _name));                                       \
+	ZBUS_RUNTIME_OBSERVERS_LIST_DECL(_CONCAT(_runtime_observers_, _name));                     \
+	FOR_EACH_NONEMPTY_TERM(_ZBUS_OBS_EXTERN, (;), _observers)                                 \
+		static const struct zbus_observer *_CONCAT(_zbus_observers_, _name)[] = {          \
```
I will fix that.

rodrigopex replied:
Done.

fabiobaltieri commented on subsys/zbus/zbus.c:
```diff
@@ -0,0 +1,210 @@
+#include <zephyr/zbus/zbus.h>
+LOG_MODULE_REGISTER(zbus, CONFIG_ZBUS_LOG_LEVEL);
+
+#if defined(CONFIG_THREAD_NAME)
+#define _ZBUS_CURRENT_THREAD_INFO k_thread_name_get(k_current_get()), k_current_get()
+#else
+#define _ZBUS_CURRENT_THREAD_INFO "_", k_current_get()
+#endif /* CONFIG_THREAD_NAME */
```
is this used anywhere?

rodrigopex replied:
No. removed.

fabiobaltieri commented on subsys/zbus/zbus_iterable_sections.c:
```diff
@@ -0,0 +1,29 @@
+bool zbus_iterate_over_channels(bool (*iterator_func)(const struct zbus_channel *chan))
+{
+	STRUCT_SECTION_FOREACH(zbus_channel, chan)
+	{
```
The bracket should go on the previous line `STRUCT_SECTION_FOREACH(zbus_channel, chan) {`

rodrigopex replied:
Fixed.

fabiobaltieri commented on subsys/zbus/zbus_iterable_sections.c:
```diff
@@ -0,0 +1,29 @@
+bool zbus_iterate_over_observers(bool (*iterator_func)(const struct zbus_observer *obs))
+{
+	STRUCT_SECTION_FOREACH(zbus_observer, obs)
+	{
```
same

rodrigopex replied:
Fixed.

fabiobaltieri commented on tests/subsys/zbus/unittests/src/main.c:
```diff
@@ -0,0 +1,565 @@
+#include <zephyr/ztest.h>
+LOG_MODULE_DECLARE(zbus, CONFIG_ZBUS_LOG_LEVEL);
+
+bool hard_msg_validator(const void *msg, size_t msg_size)
```
should be `static`

rodrigopex replied:
Fixed.

rodrigopex commented:
Thank you very much to everyone involved who gave suggestions and discussed the feature and pitfalls of the solution. I am happy to have that merged. Now let's search for bugs and fix them if they happen. Regards.

M1cha commented:
@rodrigopex How did you create those insanely pretty SVGs? :eyes:

rodrigopex commented:
> @rodrigopex How did you create those insanely pretty SVGs? 👀

Thanks @M1cha. I used Figma. The assets and images are public. You can use that by follwing the [link]( https://www.figma.com/file/0OdgYLqoPcdvlUwa7gXTfq/ZBus-Diagram-Assets-(Community)?type=design&node-id=2%3A161&mode=design&t=YvMW2giKBS5uA6F3-1). Feel free to change. Regards.
